File: sandbox/javascript/worker.js
'use strict'
/**
 * worker.js — Kafka consumer/producer worker for the JavaScript sandbox pool.
 *
 * Consumes jobs from `visualize-jobs` (filtered by language === "javascript"),
 * executes the submitted code inside a resource-limited child process
 * running tracer.js, and publishes the resulting steps[] trace to
 * `visualize-results`.
 */

const { Kafka } = require('kafkajs')
const { spawn } = require('child_process')
const path = require('path')

const KAFKA_BROKERS = (process.env.KAFKA_BROKERS || 'localhost:9092').split(',')
const JOBS_TOPIC = process.env.KAFKA_JOBS_TOPIC || 'visualize-jobs'
const RESULTS_TOPIC = process.env.KAFKA_RESULTS_TOPIC || 'visualize-results'
const GROUP_ID = process.env.KAFKA_GROUP_ID || 'sandbox-javascript'
const LANGUAGE_FILTER = process.env.LANGUAGE_FILTER || 'javascript'
const MAX_EXEC_SECONDS = parseFloat(process.env.MAX_EXEC_SECONDS || '10')
const MAX_MEMORY_MB = parseInt(process.env.MAX_MEMORY_MB || '128', 10)

const TRACER_PATH = path.join(__dirname, 'tracer.js')

function executeCode(code) {
  return new Promise((resolve) => {
    const start = Date.now()
    const child = spawn(
      process.execPath,
      [`--max-old-space-size=${MAX_MEMORY_MB}`, TRACER_PATH],
      { stdio: ['pipe', 'pipe', 'pipe'] }
    )

    let stdout = ''
    let stderr = ''
    let finished = false

    const timer = setTimeout(() => {
      if (!finished) {
        finished = true
        child.kill('SIGKILL')
        resolve({
          steps: [],
          error: `Execution timed out after ${MAX_EXEC_SECONDS}s`,
          duration_ms: Date.now() - start,
        })
      }
    }, MAX_EXEC_SECONDS * 1000)

    child.stdout.on('data', (d) => { stdout += d })
    child.stderr.on('data', (d) => { stderr += d })

    child.on('close', (codeExit) => {
      if (finished) return
      finished = true
      clearTimeout(timer)
      const duration_ms = Date.now() - start

      if (codeExit !== 0) {
        const tail = stderr.trim().split('\n').pop() || 'unknown error'
        resolve({ steps: [], error: `Sandbox error: ${tail}`, duration_ms })
        return
      }

      try {
        const payload = JSON.parse(stdout.trim().split('\n').pop())
        if (!payload.ok) {
          resolve({ steps: [], error: payload.error || 'unknown error', duration_ms })
        } else {
          resolve({ steps: payload.steps, error: '', duration_ms })
        }
      } catch (e) {
        resolve({ steps: [], error: 'Failed to parse tracer output', duration_ms })
      }
    })

    child.stdin.write(code)
    child.stdin.end()
  })
}

async function main() {
  console.log(`[sandbox-javascript] starting worker, group=${GROUP_ID}, filter=${LANGUAGE_FILTER}`)

  const kafka = new Kafka({ clientId: 'sandbox-javascript', brokers: KAFKA_BROKERS })
  const consumer = kafka.consumer({ groupId: GROUP_ID })
  const producer = kafka.producer()

  await producer.connect()
  await consumer.connect()
  await consumer.subscribe({ topic: JOBS_TOPIC, fromBeginning: false })

  console.log('[sandbox-javascript] ready, consuming jobs...')

  await consumer.run({
    eachMessage: async ({ message }) => {
      let job
      try {
        job = JSON.parse(message.value.toString())
      } catch {
        return
      }
      if (job.language !== LANGUAGE_FILTER) return

      console.log(`[sandbox-javascript] processing job ${job.id}`)
      const result = await executeCode(job.code || '')

      await producer.send({
        topic: RESULTS_TOPIC,
        messages: [{
          key: job.id,
          value: JSON.stringify({
            job_id: job.id,
            hash: job.hash || '',
            steps: result.steps,
            error: result.error,
            language: LANGUAGE_FILTER,
            duration_ms: result.duration_ms,
          }),
        }],
      })
      console.log(`[sandbox-javascript] published result for ${job.id} (${result.steps.length} steps)`)
    },
  })
}

main().catch((e) => {
  console.error('[sandbox-javascript] fatal error', e)
  process.exit(1)
})

File: sandbox/java/worker.py
"""
worker.py — Kafka consumer/producer worker for the Java sandbox pool.

Consumes jobs from `visualize-jobs` (filtered by language == "java"),
compiles + executes the submitted code inside a resource-limited subprocess
pipeline (javac -> java), and publishes the resulting steps[] trace to
`visualize-results`.
"""

import json
import os
import re
import shutil
import subprocess
import tempfile
import time
import traceback

from kafka import KafkaConsumer, KafkaProducer

from instrument import instrument

KAFKA_BROKERS = os.environ.get("KAFKA_BROKERS", "localhost:9092").split(",")
JOBS_TOPIC = os.environ.get("KAFKA_JOBS_TOPIC", "visualize-jobs")
RESULTS_TOPIC = os.environ.get("KAFKA_RESULTS_TOPIC", "visualize-results")
GROUP_ID = os.environ.get("KAFKA_GROUP_ID", "sandbox-java")
LANGUAGE_FILTER = os.environ.get("LANGUAGE_FILTER", "java")
MAX_EXEC_SECONDS = float(os.environ.get("MAX_EXEC_SECONDS", "20"))
MAX_COMPILE_SECONDS = float(os.environ.get("MAX_COMPILE_SECONDS", "10"))

PUBLIC_CLASS_RE = re.compile(r"public\s+class\s+(\w+)")


def execute_code(code: str) -> dict:
    start = time.time()

    m = PUBLIC_CLASS_RE.search(code)
    class_name = m.group(1) if m else "Solution"

    instrumented = instrument(code)

    tmp_dir = tempfile.mkdtemp(prefix="algo-java-job-")
    try:
        src_path = os.path.join(tmp_dir, f"{class_name}.java")
        with open(src_path, "w") as f:
            f.write(instrumented)

        try:
            compile_proc = subprocess.run(
                ["javac", src_path],
                capture_output=True,
                text=True,
                timeout=MAX_COMPILE_SECONDS,
                cwd=tmp_dir,
            )
        except subprocess.TimeoutExpired:
            return {"steps": [], "error": f"Compilation timed out after {MAX_COMPILE_SECONDS}s",
                     "duration_ms": (time.time() - start) * 1000}

        if compile_proc.returncode != 0:
            tail = compile_proc.stderr.strip().splitlines()
            msg = tail[-1] if tail else "compilation failed"
            return {"steps": [], "error": f"Compile error: {msg}", "duration_ms": (time.time() - start) * 1000}

        try:
            run_proc = subprocess.run(
                ["java", "-Xmx128m", "-cp", tmp_dir, class_name],
                capture_output=True,
                text=True,
                timeout=MAX_EXEC_SECONDS,
                cwd=tmp_dir,
            )
        except subprocess.TimeoutExpired:
            return {"steps": [], "error": f"Execution timed out after {MAX_EXEC_SECONDS}s",
                     "duration_ms": (time.time() - start) * 1000}

        duration_ms = (time.time() - start) * 1000

        if run_proc.returncode != 0:
            tail = run_proc.stderr.strip().splitlines()
            msg = tail[-1] if tail else "runtime error"
            return {"steps": [], "error": f"Sandbox error: {msg}", "duration_ms": duration_ms}

        marker = "__STEPS_JSON__"
        stdout = run_proc.stdout
        idx = stdout.rfind(marker)
        if idx == -1:
            return {"steps": [{"type": "done", "info": "execution completed — no array-based state changes detected"}],
                     "error": "", "duration_ms": duration_ms}

        json_part = stdout[idx + len(marker):].strip().splitlines()[0]
        try:
            steps = json.loads(json_part)
        except json.JSONDecodeError:
            return {"steps": [], "error": "Failed to parse tracer output", "duration_ms": duration_ms}

        return {"steps": steps, "error": "", "duration_ms": duration_ms}
    finally:
        shutil.rmtree(tmp_dir, ignore_errors=True)


def main():
    print(f"[sandbox-java] starting worker, group={GROUP_ID}, filter={LANGUAGE_FILTER}", flush=True)

    consumer = KafkaConsumer(
        JOBS_TOPIC,
        bootstrap_servers=KAFKA_BROKERS,
        group_id=GROUP_ID,
        value_deserializer=lambda v: json.loads(v.decode("utf-8")),
        auto_offset_reset="latest",
        enable_auto_commit=True,
    )
    producer = KafkaProducer(
        bootstrap_servers=KAFKA_BROKERS,
        value_serializer=lambda v: json.dumps(v).encode("utf-8"),
    )

    print("[sandbox-java] ready, consuming jobs...", flush=True)

    for message in consumer:
        job = message.value
        if job.get("language") != LANGUAGE_FILTER:
            continue

        job_id = job.get("id")
        code = job.get("code", "")
        print(f"[sandbox-java] processing job {job_id}", flush=True)

        try:
            result = execute_code(code)
        except Exception as e:  # noqa: BLE001
            traceback.print_exc()
            result = {"steps": [], "error": f"internal sandbox error: {e}", "duration_ms": 0}

        payload = {
            "job_id": job_id,
            "hash": job.get("hash", ""),
            "steps": result["steps"],
            "error": result["error"],
            "language": LANGUAGE_FILTER,
            "duration_ms": result["duration_ms"],
        }
        producer.send(RESULTS_TOPIC, value=payload)
        producer.flush()
        print(f"[sandbox-java] published result for {job_id} ({len(result['steps'])} steps)", flush=True)


if __name__ == "__main__":
    main()
